tests/testthat/test-scales.r

```diff
@@ -319,3 +319,58 @@ test_that("limits with NA are replaced with the min/max of the data for continuo
   expect_equal(make_scale(limits = c(1, NA), data = 1:5)$get_limits(), c(1, 5))
   expect_equal(make_scale(limits = c(NA, 5), data = 1:5)$get_limits(), c(1, 5))
 })
+
+test_that("scale_apply preserves class and attributes", {
+  df <- data_frame(
+    x = structure(c(1, 2), foo = "bar", class = c("baz", "numeric")),
+    y = c(1, 1),
+    z = c("A", "B")
+  )
+
+  # Functions to make the 'baz'-class more type stable
+  `c.baz` <- function(...) {
+    dots <- list(...)
+    attris <- attributes(dots[[1]])
+    x <- do.call("c", lapply(dots, unclass))
+    attributes(x) <- attris
+    x
+  }
+  `[.baz` <- function(x, i) {
+    attris <- attributes(x)
+    x <- unclass(x)[i]
+    attributes(x) <- attris
+    x
+  }
+
+  # These S3 methods won't work within testthat, so assign to global environment
+  assign("c.baz", `c.baz`, global_env())
+  assign("[.baz", `[.baz`, global_env())
```

Comment:
I'm not familiar with this kind of test, but it seems we can use `rlang::local_binding()` for temporal assignment on the global environment. e.g. https://github.com/r-lib/vctrs/blob/ab84679b816b903cbafa9aa042ec1c9c98d385a5/tests/testthat/helper-s3.R#L9-L11

Reply:
Ah excellent, I was hoping such a thing existed but didn't know it did!